internal/resolver/dns/dns_resolver_test.go

```diff
@@ -819,34 +816,50 @@ func TestDNSResolverExponentialBackoff(t *testing.T) {
 		for i := 0; i < 10; i++ {
 			timer := <-timerChan
 			timer.Reset(0)
-			// Propagate.
-			time.Sleep(time.Millisecond)
 		}
-		cc.m1.Lock()
-		if cc.updateStateCalls != 12 {
+		// Poll to see if DNS Resolver updated state the correct number of times, which allows time for the DNS Resolver to call
+		// ClientConn update state.
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+		for {
+			timer := time.NewTimer(time.Millisecond)
+			select {
+			case <-ctx.Done():
+				t.Fatalf("Exponential backoff is not working as expected - should update state 11 times instead of %d", cc.updateStateCalls)
+			case <-timer.C:
+			}
+			cc.m1.Lock()
+			if cc.updateStateCalls == 11 {
+				cc.m1.Unlock()
+				break
+			}
 			cc.m1.Unlock()
-			t.Fatalf("Exponential backoff is not working as expected - should be updating state at least 10 total times instead of %d", cc.updateStateCalls)
 		}
```

Comment:
This can be simplified a bit by replacing the timer & select with a sleep, since we don't need to be hyper paranoid about the context:

```go
for {
	cc.m1.Lock()
	if got := cc.updateStateCalls; got == 11 {
		cc.m1.Unlock()
		break
	}
	cc.m1.Unlock()
	if ctx.Err() != nil {
		t.Fatalf(/* use got */)
	}
	time.Sleep(time.Millisecond)
}
```

Also, note that the failure message should use the last tested value of `updateStateCalls` -- as your code is, it's actually possible to print "should update state 11 times instead of 11".

Reply:
I thought we want to avoid sleeps in tests?

Reply:
Changed context to deadline as per your other comment. Changed to this generally. Thanks.

Reply:
This sleep would be to avoid a busy-loop, so it is fine.  The sleeps to avoid are: 

1. Wait an arbitrary period of time for something to happen.  If it happens before then, we keep sleeping for the full time anyway.  If it happens 1ms after, then we lose.  So, we either accept flakes to minimize the time penalty, or pay an unnecessary time cost to eliminate flakes.

   To remedy this, we either use a signal with a timeout or do exactly this: poll in a loop with a longer deadline and a tiny sleep between polling cycles, which provides the best of both worlds.

2. Sleep because there is a timer that we're waiting on - the timer is `x`, so we sleep `x+delta`. This means we either add a potentially huge sleep in our test or inject a shorter timer.  But then we still have to deal with the question of "what is that 'delta'?", which is actually case 1 again.

   To remedy this, we replace that timer with a signal so we don't need to sleep at all.